content/post/2020-08-30-keep.source/index.Rmd

```diff
@@ -0,0 +1,130 @@
+
+[^thx]: This came to my attention thanks to [a question by Ofek Shilon on RStudio community](https://community.rstudio.com/t/keep-source-pkgs-vs-keep-source-options/69245).
+
+It is similar to CSS, JS, HTML being minified in web development to make websites load faster.
+Now sometimes you might want to keep code with its comments: 
+as an user for being able to read it locally with all its comments,
+as a developer for debugging or profiling (to have line numbers in parsed code refer to actual line numbers you can look up in your scripts).
+
+As an user installing packages, you need to look into the `keep.source.pkgs` option in `options()` that influences the behavior of package installation, or for a specific package you'd write `install.packages("rhub", INSTALL_opts = "--with-keep.source", type = "source")`.[^lookup] If you use Windows or Mac and don't write `type = "source"`, binaries might be use in which case the `keep.source.pkgs` option is ignored.
+
+[^lookup]: When viewing source code you might get a better default experience by [loading `lookup`](https://github.com/jimhester/lookup#default-printing) in your [.Rprofile](https://rstats.wtf/r-startup.html#rprofile) **if you don't use Windows, because if you use Windows and do this you won't be able to update the loaded packages with compiled code**.
+
+As a developer working on a package, you need to make sure the source is kept as is [when building the package, thanks to the `--with-keep.source` option](https://support.rstudio.com/hc/en-us/articles/205612627-Debugging-with-RStudio#debugging-in-packages), and when loading it (lucky you, the relevant `keep.source` option is `TRUE` by default in interactive sessions :tada:).
+
+
+As a developer you might also encounter the case where `R CMD check` will tell you about another switch, in an environment variable.
+It is a switch related to package installation, since `R CMD check` will install your package for checking it . 
+[See the lines below from the R source mirror](https://github.com/wch/r-source/blob/f27cbf1a52a31cd9b9676340394946a22041a4ae/src/library/tools/R/check.R#L5248-L5253):
+```r
+                        wrapLog("Information on the location(s)",
+                                "of code generating the",
+                                paste0(sQuote("Note"), "s"),
+                                "can be obtained by re-running with",
+                                "environment variable R_KEEP_PKG_SOURCE",
+                                "set to 'yes'.\n")
+```
+
+This will only work if you keep the source when building since building happens before installing, of course. :wink:
+
+Also note that there is also a way for package maintainers to [force the installation of their package to keep the source](https://stat.ethz.ch/pipermail/r-devel/2011-April/060410.html), which seems less useful?
```

Comment:
I would not say "less useful", I am pretty sure it is useful for (some of) these packages.